crates/aptos-telemetry/src/sender.rs

```diff
@@ -0,0 +1,224 @@
+        Self {
+            noise_config: node_config.identity_key().map(NoiseConfig::new),
+            token: RwLock::new(None),
+            server_public_key: Mutex::new(None),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct TelemetrySender {
+    chain_id: ChainId,
+    peer_id: PeerId,
+    client: reqwest::Client,
+    auth_context: Arc<AuthContext>,
+}
+
+impl TelemetrySender {
+    pub fn new(chain_id: ChainId, node_config: &NodeConfig) -> Self {
+        Self {
+            chain_id,
+            peer_id: node_config.peer_id().unwrap_or(PeerId::ZERO),
+            client: reqwest::Client::new(),
+            auth_context: Arc::new(AuthContext::new(node_config)),
+        }
+    }
+
+    pub async fn send_metrics(&self, event_name: String, telemetry_dump: TelemetryDump) {
+        for i in 1..MAX_RETRIES {
+            match self.post_metrics(&telemetry_dump).await {
+                Ok(_) => {
+                    debug!(
+                        "Sent telemetry event {}, data: {:?}",
+                        &event_name, &telemetry_dump
+                    );
+                    metrics::increment_telemetry_service_successes(&event_name);
+                    return;
+                }
+                Err(error) => {
+                    if error.to_string().contains("Unauthorized") && i < MAX_RETRIES {
+                        debug!("Retrying failed send telemetry event: Error: {}.", error);
+                        continue;
+                    }
+                    error!("Failed to send telemetry event: Error: {}", error);
+                    metrics::increment_telemetry_service_failures(&event_name);
+                    return;
+                }
+            }
+        }
+    }
+
+    async fn post_metrics(&self, telemetry_dump: &TelemetryDump) -> Result<(), anyhow::Error> {
+        let token = self.get_token().await?;
+
+        // Send the request and wait for a response
+        let send_result = self
+            .client
+            .post(METRICS_URL)
+            .json::<TelemetryDump>(telemetry_dump)
+            .bearer_auth(token)
+            .send()
+            .await;
+
+        // Process the response
+        match send_result {
+            Ok(response) => {
+                let status_code = response.status();
+                if status_code.is_success() {
+                    Ok(())
+                } else if status_code == StatusCode::UNAUTHORIZED {
+                    self.reset_token();
+                    Err(anyhow!("Unauthorized"))
+                } else {
+                    Err(anyhow!("Error status received: {}", status_code))
+                }
+            }
+            Err(error) => Err(anyhow!("Error sending metrics. Err: {}", error)),
+        }
+    }
+
+    async fn get_token(&self) -> Result<String, Error> {
+        // Try to read the token holding a read lock
+        let token = { self.auth_context.token.read().as_ref().cloned() };
```

Comment:
nit: unnecessary curly braces

Reply:
Are you sure the read guard would get dropped? I am getting a write lock below so I need to make sure this read lock is dropped.